core/src/rpc_subscriptions.rs

```diff
@@ -132,53 +139,49 @@ where
-    let current_ancestors = bank_forks
-        .read()
-        .unwrap()
-        .get(current_slot)
-        .unwrap()
-        .ancestors
-        .clone();
-    for (slot, _) in current_ancestors.iter() {
-        if let Some(confirmations) = r_block_commitment_cache.get_confirmation_count(*slot) {
-            confirmation_slots.entry(confirmations).or_insert(*slot);
-        }
-    }
+    let node_root = r_block_commitment_cache.root();
+    let largest_confirmed_root = r_block_commitment_cache.largest_confirmed_root();
     drop(r_block_commitment_cache);
 
     let mut notified_set: HashSet<SubscriptionId> = HashSet::new();
     if let Some(hashmap) = subscriptions.get(hashmap_key) {
-        for (sub_id, (sink, confirmations)) in hashmap.iter() {
-            let desired_slot = if *confirmations == 0 {
-                Some(&current_slot)
-            } else if *confirmations == MAX_LOCKOUT_HISTORY + 1 {
-                Some(&root)
-            } else {
-                confirmation_slots.get(confirmations)
+        for (
+            sub_id,
+            SubscriptionData {
+                sink,
+                commitment,
+                last_notified_slot,
+            },
+        ) in hashmap.iter()
+        {
+            let slot = match commitment.commitment {
+                CommitmentLevel::Max => largest_confirmed_root,
+                CommitmentLevel::Recent => current_slot,
+                CommitmentLevel::Root => node_root,
             };
-            if let Some(&slot) = desired_slot {
-                let results = {
-                    let bank_forks = bank_forks.read().unwrap();
-                    let desired_bank = bank_forks.get(slot).unwrap();
-                    bank_method(&desired_bank, hashmap_key)
-                };
-                for result in filter_results(results, root) {
-                    notifier.notify(
-                        Response {
-                            context: RpcResponseContext { slot },
-                            value: result,
-                        },
-                        sink,
-                    );
-                    notified_set.insert(sub_id.clone());
-                }
+            let results = {
+                let bank_forks = bank_forks.read().unwrap();
+                bank_forks
+                    .get(slot)
+                    .map(|desired_bank| bank_method(&desired_bank, hashmap_key))
+                    .unwrap_or_default()
+            };
+            let mut w_last_notified_slot = last_notified_slot.write().unwrap();
+            for result in filter_results(results, *w_last_notified_slot) {
+                notifier.notify(
+                    Response {
+                        context: RpcResponseContext { slot },
+                        value: result,
+                    },
+                    sink,
+                );
+                notified_set.insert(sub_id.clone());
+                *w_last_notified_slot = slot;
```

Comment:
I think we still need to set this to the last modified slot for account subscriptions

Reply:
Oh duh, yes, I agree